### ISSUE TITLE
Make `ToSqlOutput<'a>` implement `ToSql`

### ISSUE THREAD

**Issue by ncalexan** [jgallagher/rusqlite#211]
I'd like to understand why `ToSqlOutput<'a>` does not implement `ToSql`.

Over in https://github.com/mozilla/mentat, we're using `rusqlite` to implement a knowledge store on top of an embedded SQLite store.  Part of what we're doing is implementing our own value encoding: our values are strictly richer than the set of SQLite values, and therefore we encode them as a pair `(value_type_tag, sqlite_value)` to write them to the database.  (The type tag is a `TINYINT`.)  I've worked out how to use `rusqlite::Value` to _read_ a value with a dynamic type from the SQLite store, but I'm having trouble _writing_ a SQL value with a dynamic type to SQLite store.

I think my issue is specifically around minimizing copies, because I'm pretty sure I can make
```rust
to_sql(value: &mentat::Value, value_type: mentat::ValueType) -> (rusqlite::Value, i32)
```
work.  What I want, however is to make
```rust
to_sql<'a>(value: &'a mentat::Value, value_type: mentat::ValueType) -> (rusqlite::ToSqlOutput<'a>, i32)
```
work, since that should let me avoid copying potentially large strings, blobs, etc.  (By _not work_, I mean that I can't bind `&'a rusqlite::ToSqlOutput<'a>` as a parameter.)  That is, I might sometimes produce a `ValueRef::Borrowed(...)` borrowing something internal to my `mentat::Value`, potentially avoiding copying.  The obvious thing doesn't work, since `ToSqlOutput<'a>` doesn't implement `ToSql`.  It's not trivial to implement either, since there needs to be some copying, but I might be able to figure out what is appropriate in my case.

Is the missing implementation an oversight?  Is there something I am fundamentally misunderstanding about this situation?  Does it not make sense to do this in general, so that implementing a newtype sibling to `ToSqlOutput<'a>` but specialized to my situation is the way to proceed?  Thanks!

**Comment by jgallagher** [comment]
I guess I would call it an oversight, yeah. Normally (?) `ToSqlOutput` is expected to be used by `ToSql` impls; I never thought of it being one itself. If I'm understanding your case correctly, you need to convert one type (a tuple) into two SQLite columns (a value and a tag)?

It's [possible](https://github.com/jgallagher/rusqlite/pull/212) to add this impl without any copying, so I'll merge that in and push up an update once CI passes.

**Comment by rnewman** [comment]
> If I'm understanding your case correctly, you need to convert one type (a tuple) into two SQLite columns (a value and a tag)?

Yes.

If we take `15` (an `i64`) and `ValueType::Instant`, we want `v = 15` and `value_type_tag = 4`, distinguishing it from other SQL-numeric values like longs and booleans.

Similarly, `"93fe3c60-92d9-4996-9676-c03864e4f71a"` (a String) and `ValueType::UUID` might give us a compact blob or a normalized string for `v`, alongside `value_type_tag = 11`, distinguishing it in the DB from strings, URIs, or keywords.

**Comment by jgallagher** [comment]
Gotcha. That's definitely not something `ToSql` alone can handle. Is #212 enough of a change to get you going?

**Comment by rnewman** [comment]
We of course already know the tag ourselves, and can insert it like any other value, so in practice this is more about turning an arbitrary pair into a single value. If I'm reading right, I think #212 is enough, but @ncalexan will know for sure!

**Comment by jgallagher** [comment]
Whoops, github autoclosed b/c of the #212 merge (since that one does do what this issue is asking for, at least). I'll push an update to crates.io in a minute. If you need more, can you open another issue?

**Comment by jgallagher** [comment]
Published 0.9.3 which includes this impl.

**Comment by rnewman** [comment]
Thanks for the quick turnaround, @jgallagher!

**Comment by ncalexan** [comment]
> We of course already know the tag ourselves, and can insert it like any other value, so in practice this is more about turning an arbitrary pair into a single value. If I'm reading right, I think #212 is enough, but @ncalexan will know for sure!

This was, line for line, my local commit.  (I was trying to figure out good tests for this before opening a PR!)

Thanks, @jgallagher!

**Comment by jgallagher** [comment]
> This was, line for line, my local commit. (I was trying to figure out good tests for this before opening a PR!)

Glad we're on the same page!

I'm never sure how much testing is needed around trait impls like this. Pretty confident in this case that it can't be wrong if it compiles.

**Comment by ncalexan** [comment]
@jgallagher any reason to not make `ToSqlOutput` derive `Clone`?  It's sub-values do.

**Comment by jgallagher** [comment]
Nope. Should probably derive `PartialEq` and `Debug` too. Hang tight...

**Comment by jgallagher** [comment]
Published 0.9.5 which fixes this.

**Comment by ncalexan** [comment]
On Thu, Jan 26, 2017 at 7:28 PM, John Gallagher <notifications@github.com>
wrote:

> Published 0.9.5 which fixes this.
>

Thanks John, much appreciated.
